src/fairseq2/models/llama/attention.py

```diff
@@ -0,0 +1,278 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional, final
+
+from torch import Tensor
+
+from fairseq2.nn.incremental_state import IncrementalStateBag
+from fairseq2.nn.position_encoder import PositionEncoder
+from fairseq2.nn.projection import Projection
+from fairseq2.nn.transformer.attention import SDPA, create_default_sdpa
+from fairseq2.nn.transformer.multihead_attention import (
+    AttentionOutputProjection,
+    MultiheadAttention,
+    MultiheadAttentionState,
+    QKVProjection,
+)
+from fairseq2.typing import DataType, Device, finaloverride
+
+
+@final
+class GroupedQueryAttention(MultiheadAttention):
```

Comment:
I think it would be much cleaner if we could add GQA directly to `StandardMultiheadAttention`. If I am not missing anything, that would mean introducing a new `num_key_value_heads: Optional[int]` to the parameter list and updating `k_proj` and `v_proj` handling with the changes you have in this implementation.

Reply:
Added GQA to StandardMultiheadAttention in the revision.